Reconstruction/RecCalorimeter/src/components/SingleCaloTowerTool.cpp

```diff
@@ -41,96 +43,134 @@ StatusCode SingleCaloTowerTool::initialize() {
-  // Get the input collection with cells from simulation + digitisation (after calibration and with noise)
+  // Get the input collection with cells from simulation + digitisation (after
+  // calibration and with noise)
   const fcc::CaloHitCollection* cells = m_cells.get();
   debug() << "Input Hit collection size: " << cells->size() << endmsg;
   // Loop over a collection of calorimeter cells and build calo towers
-  int iPhi = 0, iEta = 0;
-  float etaCell = 0;
+  // borders of the cell in eta/phi
+  float etaCellMin, etaCellMax;
+  float phiCellMin, phiCellMax;
+  // tower index of the borders of the cell
+  int iPhiMin = 0, iPhiMax = 0;
+  int iEtaMin = 0, iEtaMax = 0;
+  // fraction of cell in eta/phi belonging a tower
+  float ratioEta = 1.0, ratioPhi = 1.0;
+  float fracEtaMin = 1.0, fracEtaMax = 1.0, fracEtaRest = 1.0;
+  float fracPhiMin = 1.0, fracPhiMax = 1.0, fracPhiRest = 1.0;
+  float epsilon = 0.0001;
   for (const auto& cell : *cells) {
-    // find to which tower the cell belongs
-    etaCell = m_segmentation->eta(cell.core().cellId);
-    iEta = idEta(etaCell);
-    iPhi = idPhi(m_segmentation->phi(cell.core().cellId));
-    // save transverse energy
-    aTowers[iEta][iPhi] += cell.core().energy / cosh(etaCell);
+    // find to which tower(s) the cell belongs
+    etaCellMin = m_segmentation->eta(cell.core().cellId) - m_segmentation->gridSizeEta() * 0.5;
+    etaCellMax = m_segmentation->eta(cell.core().cellId) + m_segmentation->gridSizeEta() * 0.5;
+    phiCellMin = m_segmentation->phi(cell.core().cellId) - M_PI / (double)m_segmentation->phiBins();
+    phiCellMax = m_segmentation->phi(cell.core().cellId) + M_PI / (double)m_segmentation->phiBins();
+    iEtaMin = idEta(etaCellMin + epsilon);
+    iPhiMin = idPhi(phiCellMin + epsilon);
+    iEtaMax = idEta(etaCellMax - epsilon);
+    iPhiMax = idPhi(phiCellMax - epsilon);
+    // if a cell is larger than a tower in eta/phi, calculate the fraction of
+    // the cell belonging to the first/last/middle tower
+    fracEtaMin = 1.0;
+    fracEtaMax = 1.0;
+    fracEtaRest = 1.0;
+    if (iEtaMin != iEtaMax) {
+      fracEtaMin = fabs(eta(iEtaMin) + 0.5 * m_deltaEtaTower - etaCellMin) / m_segmentation->gridSizeEta();
+      fracEtaMax = fabs(etaCellMax - eta(iEtaMax) + 0.5 * m_deltaEtaTower) / m_segmentation->gridSizeEta();
+      fracEtaRest = (1 - fracEtaMin - fracEtaMax) / float(iEtaMax - iEtaMin - 1);
+    }
+    fracPhiMin = 1.0;
+    fracPhiMax = 1.0;
+    fracPhiRest = 1.0;
+    if (iPhiMin != iPhiMax) {
+      fracPhiMin =
+          fabs(phi(iPhiMin) + 0.5 * m_deltaPhiTower - phiCellMin) / (2 * M_PI / (double)m_segmentation->phiBins());
+      fracPhiMax =
+          fabs(phiCellMax - phi(iPhiMax) + 0.5 * m_deltaPhiTower) / (2 * M_PI / (double)m_segmentation->phiBins());
+      if (fracPhiMin > 1 || fracPhiMax > 1) {
+        warning() << "Fraction if cekk in the tower in phi > 1 not possible!!! " << fracPhiMin << "  " << fracPhiMax
```

Comment:
typos: if -> of, cekk ->cell; add what numbers you print to make the warning more understandable?